### PR TITLE
Refactored unitary event

### PR DESCRIPTION
**Pull request by dizcza**
Seems like UE analysis is the first candidate for the new docs, DataObj, etc. I decided to take a look there. What I did:
1. Fixed docs so that smart IDEs like PyCharm highlight your code if you mess up with the input/output data types.
2. Most importantly, removed misleading `N` arg which states for the number of neurons. Currently, the user is given a false thought that he or she somehow can influence the results by modifying this parameter. However, internal logic dismisses this eventually. I see this happens in other modules too, but I started with UE.
3. Vectorized some internal loops in hashing and inverse hashing.

In a nutshell, I spent 1 day on this with a focus on increasing the readability rather than the speed. I don't think the speed is crucial here. But if it's the case, I'll open a separate PR.

### CODE REVIEW

File: elephant/unitary_event_analysis.py [hunk 1]
```diff
@@ -71,134 +87,131 @@ def hash_from_pattern(m, N, base=2):
...
-    for j, hh in enumerate(h):
-        i = N - 1
-        while i >= 0 and hh != 0:
-            m[i, j] = hh % base
-            hh /= base
-            i -= 1
+        raise ValueError("hash values should be integers")
+
+    h = np.asarray(h, dtype=np.int64)
+    powers = np.power(base, np.arange(N)[::-1])
```
Comment (dizcza) [inline elephant/unitary_event_analysis.py]: make lists

File: elephant/unitary_event_analysis.py [hunk 2]
```diff
@@ -71,134 +87,131 @@ def hash_from_pattern(m, N, base=2):
...
-               [0, 0]])
+            [1, 1],
+            [0, 1],
+            [0, 0]])
     """
 
     # check if the hash values are not greater than the greatest possible
     # value for N neurons with the given base
-    if np.any(h > np.sum([base**x for x in range(N)])):
+    if np.any(h > np.power(base, np.arange(N)).sum()):
```
Comment (dizcza) [inline elephant/unitary_event_analysis.py]: make lists

File: elephant/unitary_event_analysis.py [hunk 3]
```diff
@@ -71,134 +87,131 @@ def hash_from_pattern(m, N, base=2):
...
         raise ValueError(
             "hash value is not compatible with the number of neurons N")
     # check if the hash values are integer
     if not np.all(np.int64(h) == h):
```
Comment (dizcza) [inline elephant/unitary_event_analysis.py]: raise error if float

File: elephant/unitary_event_analysis.py [hunk 4]
```diff
@@ -312,67 +320,63 @@ def _n_exp_mat_surrogate(mat, N, pattern_hash, n_surr=1):
...
+    Parameters
     -----------
-    mat: 2d numpy array
-            the entries are zero or one
-            0-axis --> neurons
-            1-axis --> time bins
-    pattern_hash: list of integers
-            array of hash values, length: number of patterns
+    mat: np.ndarray
+         the entries are zero or one,
```
Comment (dizcza) [inline elephant/unitary_event_analysis.py]: change the docs

File: elephant/unitary_event_analysis.py [hunk 5]
```diff
@@ -576,14 +561,14 @@ def jointJ(p_val):
     for better visualization as the highly significant events are
     indicated by very low joint-p-values
 
-    Parameters:
+    Parameters
     -----------
-    p_val: list of floats
-        p-values of statistical tests for different pattern.
+    p_val: list
+        List of p-values of statistical tests for different pattern.
```
Comment (dizcza) [inline elephant/unitary_event_analysis.py]: list of floats

File: elephant/unitary_event_analysis.py [hunk 6]
```diff
@@ -576,14 +561,14 @@ def jointJ(p_val):
...
 
-    Returns:
+    Returns
     --------
-    J: list of floats
+    Js: list
```
Comment (dizcza) [inline elephant/unitary_event_analysis.py]: list of loats

**Comment by dizcza** [comment]
add a warning if `N` is not None -> deprecation warning
